stdlib/public/Atomics/UnsafeAtomic.swift.gyb

```diff
@@ -0,0 +1,376 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+%{
+  def lowerFirst(str):
+    return str[:1].lower() + str[1:] if str else ""
+
+  def argLabel(label):
+    return label + ": " if label <> "_" else ""
+
+  integerOperations = [
+    # Swift name,         llvm name,  operator, label,  doc
+    ('WrappingIncrement', 'add',      '&+',     "by",   "wrapping add"),
+    ('WrappingDecrement', 'sub',      '&-',     "by",   "wrapping subtract"),
+    ('BitwiseAnd',        'and',      '&',      "with", "bitwise and"),
+    ('BitwiseOr',         'or',       '|',      "with", "bitwise or"),
+    ('BitwiseXor',        'xor',      '^',      "with", "bitwise xor")
+  ]
+
+  def argLabel(label):
+    return label + ": " if label <> "_" else ""
+
+  def defaultValue(name):
+    return " = 1" if "crement" in name else ""
+}%
+
+import Swift
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+@frozen
+public struct UnsafeAtomic<Value: AtomicProtocol> {
+  @frozen
+  public struct Storage {
+    @usableFromInline
+    internal var _value: Value._AtomicStorage
+
+    @inlinable @inline(__always)
+    internal init(_value: __owned Value._AtomicStorage) {
+      self._value = _value
+    }
+
+    @inlinable @inline(__always)
+    public init(_ value: __owned Value) {
+      self._value = Value._prepareAtomicStorage(for: value)
+    }
+
+    @inlinable @inline(__always)
+    @discardableResult
+    public mutating func dispose() -> Value {
+      return Value._disposeAtomicStorage(&_value)
+    }
+  }
```

Comment:
Again, this is technically possible, but I prefer to have a dedicated, public wrapper so that particular choices for `_AtomicStorage` won't get hardwired into (ostensibly) portable codebases.

`UnsafeAtomic.Storage` has the additional benefit that we can attach API to it (`init(_:)`, `dispose()`) that would be inappropriate on `_AtomicStorage`, esp. in the case where atomic types use `Self` as their atomic storage.